unison-cli/src/Unison/Codebase/Editor/HandleInput.hs

```diff
@@ -2000,6 +1945,81 @@ handleShowDefinition outputLoc inputQuery = do
+handleTest :: Monad m => TestInput -> Action' m v ()
+handleTest TestInput {includeLibNamespace, showFailures, showSuccesses} = do
+  testTerms <- do
+    currentPath' <- use LoopState.currentPath
+    currentBranch' <- getAt currentPath'
+    currentBranch'
+      & Branch.head
+      & Branch.deepTermMetadata
+      & R4.restrict34d12 isTest
+      & (if includeLibNamespace then id else R.filterRan (not . isInLibNamespace))
+      & R.dom
+      & pure
+  let testRefs = Set.mapMaybe Referent.toTermReference testTerms
+      oks results =
+        [ (r, msg)
+          | (r, Term.List' ts) <- Map.toList results,
+            Term.App' (Term.Constructor' (ConstructorReference ref cid)) (Term.Text' msg) <- toList ts,
+            cid == DD.okConstructorId && ref == DD.testResultRef
+        ]
+      fails results =
+        [ (r, msg)
+          | (r, Term.List' ts) <- Map.toList results,
+            Term.App' (Term.Constructor' (ConstructorReference ref cid)) (Term.Text' msg) <- toList ts,
+            cid == DD.failConstructorId && ref == DD.testResultRef
+        ]
+  cachedTests <- fmap Map.fromList . eval $ LoadWatches WK.TestWatch testRefs
+  let stats = Output.CachedTests (Set.size testRefs) (Map.size cachedTests)
+  names <-
+    makePrintNamesFromLabeled' $
+      LD.referents testTerms
+        <> LD.referents [DD.okConstructorReferent, DD.failConstructorReferent]
+  ppe <- fqnPPE names
+  respond $
+    TestResults
+      stats
+      ppe
+      showSuccesses
+      showFailures
+      (oks cachedTests)
+      (fails cachedTests)
+  let toCompute = Set.difference testRefs (Map.keysSet cachedTests)
+  unless (Set.null toCompute) do
+    let total = Set.size toCompute
+    computedTests <- fmap join . for (toList toCompute `zip` [1 ..]) $ \(r, n) ->
+      case r of
+        Reference.DerivedId rid -> do
+          tm <- eval $ LoadTerm rid
+          case tm of
+            Nothing -> do
+              hqLength <- eval CodebaseHashLength
+              respond (TermNotFound' . SH.take hqLength . Reference.toShortHash $ Reference.DerivedId rid)
+              pure []
+            Just tm -> do
+              respond $ TestIncrementalOutputStart ppe (n, total) r tm
+              --                          v don't cache; test cache populated below
+              tm' <- eval $ Evaluate1 ppe False tm
+              case tm' of
+                Left e -> respond (EvaluationFailure e) $> []
+                Right tm' -> do
+                  -- After evaluation, cache the result of the test
+                  eval $ PutWatch WK.TestWatch rid tm'
+                  respond $ TestIncrementalOutputEnd ppe (n, total) r tm'
+                  pure [(r, tm')]
+        r -> error $ "unpossible, tests can't be builtins: " <> show r
+
+    let m = Map.fromList computedTests
+    respond $ TestResults Output.NewlyComputed ppe showSuccesses showFailures (oks m) (fails m)
+  where
+    isInLibNamespace :: Name -> Bool
+    isInLibNamespace name =
```

Comment:
Does this work for tests in child namespaces? E.g. if I'm at the root and run `test` it seems like this wouldn't filter out tests in `base.lib.*`;

Should this maybe be `"lib" `elem` init (Name.segments name)` or something?

Reply:
The convention can really be whatever we want, but I had thought it was just "treat `lib` as a namespace that contains dependencies" and we put the responsibility on the user to `cd` to their project.